### PR TITLE
Allow users to see confirm screen errors.

### PR DESCRIPTION
**Pull request by danjm**
This PR makes it possible to actually view transaction errors that occur upon submission. Currently, when on the confirm screen and attempting to submit a transaction, an error is rendered for a half second and then the user is taken to the home screen. This leaves the user with no information about why the transaction submission attempt failed, making it harder for them to correct their own actions, and adding friction to the support process.

This PR corrects this keeping a record of failed transactions that we wish to show, and then utilizing the confirm screen to show them. When in this state, transaction data cannot be edited. For now, this is just utilized to show failed transactions immediately after they fail. In the future we could add the ability to inspect transactions that have failed in the past via the same UI, but we don't need to.

This was motivated by developers and users of custom networks that have varying gas limits. It was suggested that allowing users to see this error would help them adapt to the custom network more easily.

This is meant to be an iterative but clear improvement over the current state, which simply allows a user to actually read an error that we render but navigate away from too quickly.

### Before

https://user-images.githubusercontent.com/7499938/123792053-c1b03780-d8ba-11eb-831c-5b2e4a0cf298.mp4

### After:

https://user-images.githubusercontent.com/44811/137756296-fba57419-b4fc-4e6f-aa42-f61475a765f1.mov

### CODE REVIEW

**Comment by danjm** [comment]
closing for now, will re-open once there are proper designs

**Comment by kevinghim** [comment]
Reopening with Yanrong's new designs: https://www.figma.com/file/vAvkQ7sbGlTCFqffCHfIfY/confirm-screen-errors?node-id=0%3A1

**Comment by danjm** [comment]
Once resolved, this PR will also address: https://github.com/MetaMask/metamask-extension/issues/11394

**Comment by danjm** [comment]
It might be best to close this draft PR and just reference it in the ticket.

**Comment by segun** [comment]
I'm looking to fix this, but I can't seem to find where to pause the process from going back to the home page when there's an error. It sort of auto switched to the home page once the button is clicked.

@danjm @brad-decker

**Comment by darkwing** [comment]
Something weird happened here with the number of files changed...

**Comment by segun** [comment]
> Something weird happened here with the number of files changed...

ooops, it was 21. I did a rebase. I think that funked things up.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

**Review by jpuri — COMMENTED** [review]
Nice work 👍 , I added some small feedbacks.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: ui/store/actions.js [hunk 1]
```diff
@@ -1103,6 +1113,32 @@ export function updateMetamaskState(newState) {
       type: actionConstants.UPDATE_METAMASK_STATE,
       value: newState,
     });
+
+    // check that tx status has not changed and remove it from failed transactions if it has
+    const foundTx = currentNetworkTxList && currentNetworkTxList.find((currentTx, index) => {
```
Comment (Gudahtt) [inline ui/store/actions.js:1118]: The `currentNetworkTxList` should always be set. If it's empty, it'll be an empty array.

```suggestion
    const foundTx = currentNetworkTxList.find((currentTx, index) => {
```

Comment (Gudahtt) [reply ui/store/actions.js:1118]: Also... this only finds the first transaction in `currentNetworkTxList` that should be removed from `failedTransactionsToDisplay`. If multiple transactions were to be removed in one update, it'd miss the second one.

I think it would be better to loop through `failedTransactionsToDisplay` and check each one against `currentNetworkTxList` for state changes. That way we ensure we remove _all_ that should be removed. 

e.g. `const transactionIdsToRemove = failedTransactionsToDisplay.filter(...)`

**Review by Gudahtt — COMMENTED** [review]
The name `failedTxesToDisplay` seems to wrongly apply that all of the transactions in the list have failed already. But actually this is a list of transactions that _might_ encounter a network failure, which we want to display in the event of an error. Maybe we can come up with a more accurate name.

And on a less important note, I think it would be better for readability to spell out works completely rather than use shortened forms like "txs".

`transactionsToDisplayUponFailure` maybe? I'm having trouble thinking of alternate names.

File: ui/store/actions.js [hunk 2]
```diff
@@ -1103,6 +1113,32 @@ export function updateMetamaskState(newState) {
...
+      if (
+        failedTransactionsToDisplay &&
+        failedTransactionsToDisplay[currentTx.id]
+      ) {
+        const newTxStatus = getStatusKey(newTx);
+        if (
+          isEqual(newTx.status, currentTx.status) === false &&
+          newTxStatus !== TRANSACTION_STATUSES.FAILED &&
+          newTxStatus !== TRANSACTION_STATUSES.SIGNED &&
+          newTxStatus === TRANSACTION_STATUSES.APPROVED
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/store/actions.js]: It looks like this hasn't been addressed yet

File: ui/store/actions.js [hunk 3]
```diff
@@ -1116,6 +1126,37 @@ export function updateMetamaskState(newState) {
       type: actionConstants.UPDATE_METAMASK_STATE,
       value: newState,
     });
+
+    // Check that the transaction was not submitted successfully, and remove it from failed transactions if it was.
+    const transactionIdsToRemove = Object.values(
```
Comment (Gudahtt) [inline ui/store/actions.js:1131]: Was this meant to be `Object.keys`? The values are all booleans, not ids.

File: ui/components/ui/error-message/index.scss [hunk 4]
```diff
@@ -19,4 +19,18 @@
   &__text {
     overflow: auto;
   }
+
+  &__link {
+    text-decoration: none;
+    cursor: pointer;
```
Comment (Gudahtt) [inline ui/components/ui/error-message/index.scss:25]: Now that the link is an `<a>` element, it already has a pointer cursor. This is not needed.

```suggestion
```

File: ui/components/ui/error-message/error-message.component.js [hunk 5]
```diff
@@ -12,14 +12,25 @@ const ErrorMessage = (props, context) => {
         alt=""
         className="error-message__icon"
       />
-      <div className="error-message__text">{error}</div>
+      <div className="error-message__text">
+        {error}
+        {linkText ? (
+          <a className="error-message__link" onClick={onErrorMessageClick}>
```
Comment (Gudahtt) [inline ui/components/ui/error-message/error-message.component.js:18]: IIRC this needs an `href` to be keyboard focusable. Elsehwere we've been settting `href` to `"#"` if we don't want to navigate anywhere, just to ensure it's focusable.

```suggestion
          <a className="error-message__link" href="#" onClick={onErrorMessageClick}>
```

File: ui/store/actions.js (hunk 3 above)
Comment (segun) [reply ui/store/actions.js:1131]: yes object.keys.

File: ui/store/actions.js [hunk 6]
```diff
@@ -1116,6 +1126,37 @@ export function updateMetamaskState(newState) {
       type: actionConstants.UPDATE_METAMASK_STATE,
       value: newState,
     });
+
+    // Check that the transaction was not submitted successfully, and remove it from failed transactions if it was.
+    const transactionIdsToRemove = Object.keys(
+      transactionsToDisplayOnFailure,
+    ).filter((id) => {
+      const foundTx = currentNetworkTxList.find((currentTx, index) => {
```
Comment (Gudahtt) [inline ui/store/actions.js:1134]: I don't think this is doing what you meant it to do. It looks like this is finding the first transaction in `currentNetworkTxList` that has changed to a relevant status, irrespective of the id. Then it's checking the ID last. This will find at-most a single transaction.

File: ui/store/actions.js [hunk 7]
```diff
@@ -1116,6 +1126,39 @@ export function updateMetamaskState(newState) {
...
+        .map((tx, index) => {
+          return { id: tx.id, status: tx.status, index };
+        })
+        .find((currentTx) => {
+          return id === currentTx.id;
+        });
+
+      // check if the status of foundTx has changed
+      if (foundTx) {
+        const newTx = newNetworkTxList[foundTx.index];
```
Comment (Gudahtt) [inline ui/store/actions.js:1144]: There is no guarantee that the transaction at this index is the correct transaction. We need to find this transaction by id.

File: ui/store/actions.js [hunk 8]
```diff
@@ -1116,6 +1126,39 @@ export function updateMetamaskState(newState) {
...
+          newTx.status !== foundTx.status &&
+          newTxStatus !== TRANSACTION_STATUSES.FAILED &&
+          newTxStatus !== TRANSACTION_STATUSES.SIGNED &&
+          newTxStatus !== TRANSACTION_STATUSES.APPROVED
+        ) {
+          return true;
+        }
+      }
+
+      return false;
```
Comment (Gudahtt) [inline ui/store/actions.js:1156]: This can be done more concisely, e.g.

```suggestion
      const currentTx = currentNetworkTxList.find((tx) => tx.id === id);
      const newTx = newNetworkTxList.find((tx) => tx.id === id);

      return (
        newTx.status !== currentTx.status &&
        newTx.status !== TRANSACTION_STATUSES.FAILED &&
        newTx.status !== TRANSACTION_STATUSES.SIGNED &&
        newTx.status !== TRANSACTION_STATUSES.APPROVED
      );
```

This also solves the issue I mentioned in my other comment about how to find the transaction in the `newNetworkTxList`.

File: ui/selectors/selectors.js [hunk 9]
```diff
@@ -403,6 +403,28 @@ export function getUnapprovedConfirmations(state) {
...
 
+export function getFailedTransactionsToDisplay(state) {
+  const { currentNetworkTxList = [] } = state.metamask;
+  const { transactionsToDisplayOnFailure = {} } = state.appState;
+  const failedTransactions = currentNetworkTxList.reduce(
+    (_failedTransactions, tx) => {
+      if (
+        transactionsToDisplayOnFailure[tx.id] &&
+        (tx.err || tx.status === 'approved' || tx.status === 'signed')
+      ) {
```
Comment (Gudahtt) [inline ui/selectors/selectors.js:414]: I don't think we need to filter transactions by state or by the presence of an error here. If a transaction is in the `transactionsToDisplayOnFailure` map, we should include it here. The logic for filtering out completed transactions is already present in the background state update handler - we don't need to duplicate it.

This is dangerous in that it could hide an error if we did include a transaction by mistake that doesn't belong here. Also it would result in a mismatch between this list and the "display count" derived below.

```suggestion
      if (transactionsToDisplayOnFailure[tx.id]) {
```

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js [hunk 10]
```diff
@@ -84,10 +89,13 @@ const mapStateToProps = (state, ownProps) => {
...
     provider: { chainId },
   } = metamask;
+
+  const transactionsToDisplayOnFailure = getFailedTransactionsToDisplay(state);
+
   const { tokenData, txData, tokenProps, nonce } = confirmTransaction;
   const { txParams = {}, id: transactionId, type } = txData;
   const transaction =
-    Object.values(unapprovedTxs).find(
+    Object.values({ ...unapprovedTxs, ...transactionsToDisplayOnFailure }).find(
```
Comment (Gudahtt) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:98]: In this case we want to ensure that the unapproved transaction is always chosen first, if it's present in both collections. It probably works correctly here because `Object.values` returns entries in insert order in most cases. But I do not think the iteration order of `Object.values` is guaranteed to be insert-order in all cases. I don't know if this works consistently in all of our supported browsers.

To make certain that this check is done in the correct order, we should look in each of these collections separately:

e.g. something like:

```
const transaction =
  Object.values(unapprovedTxs).find(({ id }) => id === (transactionId || Number(paramsTransactionId))) ||
  Object.values(transactionsToDisplayOnFailure).find(({ id }) => id === (transactionId || Number(paramsTransactionId))) ||
  {};
```

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js [hunk 11]
```diff
@@ -180,6 +188,10 @@ const mapStateToProps = (state, ownProps) => {
     fromAddress,
   );
 
+  const isFailedTransaction = Boolean(
+    transactionsToDisplayOnFailure[fullTxData.id],
```
Comment (Gudahtt) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:192]: The presence of the transaction in this collection does not indicate that it is a failed transaction. We should be checking here that the transaction _isn't_ in `unapprovedTxs` - not that it _is_ in `transactionsToDisplayOnFailure`.

File: ui/components/app/confirm-page-container/confirm-page-container-container.test.js [hunk 12]
```diff
@@ -61,6 +61,7 @@ describe('Confirm Page Container Container Test', () => {
     showAddToAddressBookModal: sinon.spy(),
     contact: undefined,
     isOwnedAccount: false,
+    hideTitle: false,
```
Comment (Gudahtt) [inline ui/components/app/confirm-page-container/confirm-page-container-container.test.js:64]: Why was this change needed? It appears to be unrelated to the other changes.

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js [hunk 13]
```diff
@@ -117,15 +124,32 @@ export default class ConfirmPageContainerContent extends Component {
         {this.renderContent()}
         {(errorKey || errorMessage) && (
           <div className="confirm-page-container-content__error-container">
-            <ErrorMessage errorMessage={errorMessage} errorKey={errorKey} />
+            <ErrorMessage
+              errorMessage={this.context.t('somethingWentWrong')}
+              errorKey={errorKey}
```
Comment (Gudahtt) [inline ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:129]: In the case where `errorKey` is set, this component is broken.

`errorKey`, when set, is used to obtain the error message. This is a deprecated pattern that we're trying to get rid of because it makes these keys impossible to find with strict-mode of the localized message linter, but nonetheless it's something we need to continue to support until we remove all uses.

Right now if it's set, the full error message will be shown (instead of "something went wrong") along with the "more details" button. When the user clicks "more details", they'll be shown an empty modal.

We need to decide - do we want to treat keyed errors the same way we do text errors? If so, we need to stop passing it into `ErrorMessage` here, and use the key to get the full error message below when passing it into `TransactionErrorDetailsModal`.

If we want to treat keyed errors differently than raw string errors (because they have translations), we can show them directly here and omit the "More details" link altogether.

File: ui/components/app/gas-customization/advanced-gas-inputs/advanced-gas-inputs.component.js [hunk 14]
```diff
@@ -21,6 +21,7 @@ export default class AdvancedGasInputs extends Component {
     minimumGasLimit: PropTypes.number,
     customPriceIsExcessive: PropTypes.bool,
     networkSupportsSettingGasPrice: PropTypes.bool,
+    disabled: PropTypes.bool,
```
Comment (Gudahtt) [inline ui/components/app/gas-customization/advanced-gas-inputs/advanced-gas-inputs.component.js:24]: Why were the changes in this component made? It looks like this prop is never set.

Comment (segun) [reply ui/components/app/gas-customization/advanced-gas-inputs/advanced-gas-inputs.component.js:24]: These crept in after the last rebase....I am going to compare with previous commits and fix.

File: ui/components/app/confirm-page-container/confirm-page-container-container.test.js (hunk 12 above)
Comment (segun) [reply ui/components/app/confirm-page-container/confirm-page-container-container.test.js:64]: This snuck in from the last rebase.

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js (hunk 13 above)
Comment (segun) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:129]: @Gudahtt What do you about about splitting this component, so we have one with the ability to accept links and the other one without?

Comment (Gudahtt) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:129]: That could work! They seem different enough - one with plain text (the error key case) and one with a generic error with a more details button. Good idea.

Comment (segun) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:129]: cool

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js (hunk 11 above)
Comment (segun) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js]: @Gudahtt this doesn't work as expected. At this point in the execution, the transaction is in unapprovedTxs even though it's a failed tx.

Comment (Gudahtt) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:201]: Hmmm. That should not be the case. The list of `unapprovedTxs` necessarily have a status of "unapproved".

File: ui/components/ui/detailed-error-message/detailed-error-message.component.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+          ''
+        )}
+      </div>
+    </div>
+  );
+};
+
+DetailedErrorMessage.propTypes = {
+  errorMessage: PropTypes.string,
+  errorKey: PropTypes.string,
```
Comment (Gudahtt) [inline ui/components/ui/detailed-error-message/detailed-error-message.component.js:31]: I think we should omit this prop, since we want to discourage the use of "error keys" (because they are incompatible with the strict mode of our verify-locales linter) and none are currently used with this component.

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js (hunk 11 above)
Comment (segun) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:201]: When I use the above condition (isFailedTransaction = isn't in unapprovedTxs), I can't go past the confirm page because the confirm button is not shown (i.e isFailedTransaction is true).

Pretty late for me. I will take a closer look in the morning

File: ui/components/ui/detailed-error-message/detailed-error-message.component.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+  linkText: PropTypes.string,
+  onErrorMessageClick: PropTypes.func,
```
Comment (Gudahtt) [inline ui/components/ui/detailed-error-message/detailed-error-message.component.js:33]: Could we make these mandatory? I can't imagine we'd have a "detailed error message" that had no way to click to see the details.
```suggestion
  linkText: PropTypes.string.isRequired,
  onErrorMessageClick: PropTypes.func.isRequired,
```

This would let us simplify the render function as well, by removing the conditions that check if these props are present or not. And it would help validate that `linkText` is always set with `onErrorMessageClick`.

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js (hunk 11 above)
Comment (Gudahtt) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js:201]: Interesting! Sounds good, I can take a look tomorrow as well

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js (hunk 13 above)
Comment (Gudahtt) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js]: The second new component looks great!

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js (hunk 11 above)
Comment (segun) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js]: Ok, so I fixed this. The real issue was fullTxData.id is number and unapproveTxs ids are string.

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js [hunk 17]
```diff
@@ -117,15 +125,38 @@ export default class ConfirmPageContainerContent extends Component {
         {this.renderContent()}
         {(errorKey || errorMessage) && (
           <div className="confirm-page-container-content__error-container">
-            <ErrorMessage errorMessage={errorMessage} errorKey={errorKey} />
+            {errorKey ? (
+              <ErrorMessage
+                errorMessage={this.context.t('somethingWentWrong')}
```
Comment (Gudahtt) [inline ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:130]: This prop is not used if `errorKey` is passed in

```suggestion
```

File: ui/pages/confirm-transaction-base/confirm-transaction-base.container.js (hunk 11 above)
Comment (Gudahtt) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js]: This still looks like the opposite of what it should be. Did you forget a `!`?

Comment (Gudahtt) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.container.js]: Did you discover what was going on here after?

**Comment by danjm** [comment]
@segun can you address this previous comment:

> The name failedTxesToDisplay seems to wrongly apply that all of the transactions in the list have failed already. But actually this is a list of transactions that might encounter a network failure, which we want to display in the event of an error. Maybe we can come up with a more accurate name.

> transactionsToDisplayUponFailure maybe? I'm having trouble thinking of alternate names.

**Comment by danjm** [comment]
Oh, I see that was done at some point: https://github.com/MetaMask/metamask-extension/commit/ac896b3f871fe530ea46745486dba0d78fbb2cec

**Comment by danjm** [comment]
Oh, it is just some methods that have names similar to the old `failedTxesToDisplay`

File: ui/pages/confirm-transaction/confirm-transaction.container.js [hunk 18]
```diff
@@ -28,10 +36,24 @@ const mapStateToProps = (state, ownProps) => {
...
-    ? unapprovedTxs[id] || unconfirmedTransactions[0]
-    : {};
+
+  if (failedTransactionsToDisplay && failedTransactionsToDisplay[id]) {
+    const status = getStatusKey(failedTransactionsToDisplay[id]);
+    if (
+      status !== TRANSACTION_STATUSES.FAILED &&
+      status !== TRANSACTION_STATUSES.SIGNED
+    ) {
+      delete failedTransactionsToDisplay[id];
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/pages/confirm-transaction/confirm-transaction.container.js]: This was addressed in https://github.com/MetaMask/metamask-extension/commit/f2345d2a62e779732c1047a20977ec82ec0eabeb

File: ui/store/actions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (danjm) [reply ui/store/actions.js]: addressed in https://github.com/MetaMask/metamask-extension/commit/9f1be122cb88e682da264b837b14f6b57717a6a3#diff-3c58b916cefff81e4927a9afd392acadbf4515f99f41783964a9851b7774791b

File: ui/store/actions.js (hunk 1 above)
Comment (danjm) [reply ui/store/actions.js]: this was done in https://github.com/MetaMask/metamask-extension/commit/ac896b3f871fe530ea46745486dba0d78fbb2cec#diff-3c58b916cefff81e4927a9afd392acadbf4515f99f41783964a9851b7774791b

**Comment by danjm** [comment]
Looks like this comment needs resolution https://github.com/MetaMask/metamask-extension/pull/11416#discussion_r756946798

**Comment by Gudahtt** [comment]
The e2e test error does seem like a legitimate issue. I tried locally and it also fails. Infinite loading screen, no console errors in UI or background.

**Review by danjm — COMMENTED** [review]
LGTM

**Comment by danjm** [comment]
Explanation of last commit: `Object.keys` always returns an array of strings, even if the keys are numbers.

**Comment by danjm** [comment]
Rebased again to fix the latest merge conflict

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 19]
```diff
@@ -1088,7 +1106,12 @@ export default class ConfirmTransactionBase extends Component {
           onEdit={() => this.handleEdit()}
           onCancelAll={() => this.handleCancelAll()}
           onCancel={() => this.handleCancel()}
-          onSubmit={() => this.handleSubmit()}
+          onSubmit={() =>
+            isFailedTransaction
+              ? this.handleFailedTxClose()
+              : this.handleSubmit()
+          }
+          onConfirmAnyways={() => this.handleConfirmAnyways()}
```
Comment (Gudahtt) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1114]: What is this for? It doesn't appear to be used.

```suggestion
```

File: ui/store/actions.js [hunk 20]
```diff
@@ -1105,6 +1112,28 @@ export function updateMetamaskState(newState) {
       type: actionConstants.UPDATE_METAMASK_STATE,
       value: newState,
     });
+
+    // Check that the transaction was not submitted successfully, and remove it from failed transactions if it was.
+
```
Comment (Gudahtt) [inline ui/store/actions.js:1117]: Nit: This wording is a bit awkward (my bad, I think this is from an earlier suggestion of mine), and this newline isn't needed. Here is a suggested improvement.

 ```suggestion
    // Remove successful transactions from `transactionsToDisplayOnFailure`.
```

File: ui/ducks/app/app.js [hunk 21]
```diff
@@ -388,6 +405,19 @@ export function hideWhatsNewPopup() {
 export function toggleGasLoadingAnimation(value) {
   return { type: actionConstants.TOGGLE_GAS_LOADING_ANIMATION, value };
 }
+export function addTxToFailedTxesToDisplay(txId) {
```
Comment (Gudahtt) [inline ui/ducks/app/app.js:408]: Could we rename these two functions and the actions as well? This isn't "failed transactions to display" anymore, it's "transactions to display on failure".

So this could be `ADD_TRANSACTION_TO_DISPLAY_ON_FAILURE` and `addTransactionToDisplayOnFailure` instead for example.

File: ui/ducks/app/app.js [hunk 22]
```diff
@@ -388,6 +405,19 @@ export function hideWhatsNewPopup() {
...
   return { type: actionConstants.TOGGLE_GAS_LOADING_ANIMATION, value };
 }
+export function addTransactionToDisplayOnFailure(txId) {
+  return {
+    type: actionConstants.ADD_TRANSACTION_TO_DISPLAY_ON_FAILURE,
+    value: txId,
+  };
+}
+
+export function removeTxFromFailedTxesToDisplay(txId) {
```
Comment (Gudahtt) [inline ui/ducks/app/app.js:415]: Sorry I had meant that both functions should be renamed. This function should be renamed too, for the same reason as `addTransactionToDisplayOnFailure`.

Comment (segun) [reply ui/ducks/app/app.js:415]: I'm very bad with names...so this is what I come up with

`removeTransactionsDisplayedOnFailure` and `REMOVE_TRANSACTIONS_DISPLAYED_ON_FAILURE`

Comment (Gudahtt) [reply ui/ducks/app/app.js]: Hmm.... to be the obvious choice would be to match the parallel "add" function, but replace "add" with "remove". e.g.

`REMOVE_TRANSACTION_TO_DISPLAY_ON_FAILURE` and `removeTransactionToDisplayOnFailure`

File: ui/components/app/confirm-page-container/confirm-page-container.component.js [hunk 23]
```diff
@@ -85,6 +85,7 @@ export default class ConfirmPageContainer extends Component {
     contact: PropTypes.object,
     isOwnedAccount: PropTypes.bool,
     supportsEIP1559V2: PropTypes.bool,
+    isFailedTransaction: PropTypes.bool,
```
Comment (Gudahtt) [inline ui/components/app/confirm-page-container/confirm-page-container.component.js:88]: The prop types for this component are organized into categories. This prop should be in the "Footer" category, but it's currently listed under "Gas Popover".

File: ui/components/app/confirm-page-container/confirm-page-container.component.js [hunk 24]
```diff
@@ -240,8 +247,14 @@ export default class ConfirmPageContainer extends Component {
...
-              submitText={this.context.t('confirm')}
+              submitText={
+                isFailedTransaction
+                  ? this.context.t('close')
+                  : this.context.t('confirm')
+              }
+              submitButtonType={isFailedTransaction ? 'default' : 'confirm'}
               disabled={disabled}
             >
               {unapprovedTxCount > 1 && (
```
Comment (Gudahtt) [inline ui/components/app/confirm-page-container/confirm-page-container.component.js:260]: Should we hide the "Reject X" button as well when showing a failed transaction? It would be a bit confusing if that button didn't reject what is currently shown.

**Review by Gudahtt — COMMENTED** [review]
I noticed that this breaks our display of multiple transactions. The code for showing the "X of Y" header when displaying multiple transactions doesn't count failed transactions, and neither does the "Reject X" button at the bottom.

I'm not entire sure what to do here design wise. We can update the counter to include a failed transaction that we want to display. But should we still show the "Reject X" button when displaying a failed transaction? Should the "X" include the transaction being shown or not (since we can't really "reject" it because it has already failed).

File: ui/components/app/confirm-page-container/confirm-page-container.component.js (hunk 24 above)
Comment (segun) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:260]: The "Reject X" button is not shown.

Comment (Gudahtt) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:260]: It is shown of there is more than one unapproved transaction. It is possible for that to be the case while viewing a failed transaction.

**Review by danjm — COMMENTED** [review]
This looks good to me. But I can't approve because I am the author of the PR

File: ui/components/app/modals/transaction-error-details-modal/index.scss [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+      font-weight: bold;
+
+      h2 {
+        white-space: nowrap;
+        overflow: hidden;
+        text-overflow: ellipsis;
+      }
+
+      button {
+        margin-right: 24px;
```
Comment (darkwing) [inline ui/components/app/modals/transaction-error-details-modal/index.scss:35]: This might be best served as `margin-inline-end` to better support RTL.

File: ui/components/ui/error-message/index.scss [hunk 26]
```diff
@@ -19,4 +19,17 @@
...
+    color: $Blue-500;
+
+    &:hover {
+      color: $Blue-400;
+    }
+
+    &:active {
+      color: $Blue-600;
+    }
+  }
```
Comment (darkwing) [inline ui/components/ui/error-message/index.scss:34]: Since we moved to CSS variables, we need to change `$Blue-500` to `var(--Blue-500)`, as with the other two.

**Review by darkwing — CHANGES_REQUESTED** [review]
ErrorDetails

The spacing of the icon and the text seems slim, and we should probably make the text size a bit smaller.  The other thing I'm seeing is that "More Details" text shows the text cursor upon hover, where it should show a pointer cursor.

File: ui/components/ui/detailed-error-message/detailed-error-message.component.js [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+  return (
+    <div className="error-message">
+      <img
+        src="./images/alert-red.svg"
+        alt=""
+        className="error-message__icon"
+      />
+      <div className="error-message__text">
+        {errorMessage}
+        <a className="error-message__link" onClick={onErrorMessageClick}>
```
Comment (darkwing) [inline ui/components/ui/detailed-error-message/detailed-error-message.component.js:16]: We should make this a `<Button type="link">` component for the sake of accessibility.

**Comment by darkwing** [comment]
Functionally this works great!  Just a few nits for CSS and accessibility improvements.

**Comment by brad-decker** [comment]
@danjm is this something we still want? Possible to get this rebased so we can get it landed?

**Comment by brad-decker** [comment]
@danjm closing this PR feel free to reopen it if you think it should remain open. in the meantime the branch will stay put if we want to return to this.
